### PR TITLE
Allow ImageCreator url from freesites

### PR DESCRIPTION
**Pull request by Olezha**
This page looks like image placeholder generator for freesites. If so, then it is worth to be able to use this URL from the site.

### CODE REVIEW

**Comment by ArneBab** [comment]
Could you merge next into your branch? There’s a fix for our Travis CI in next now.

**Comment by ArneBab** [comment]
This looks reasonable in itself, but the image creator does not look robust enough to be opened to Freesites.

You could pass it Integer.MAX_VALUE as width and height which would likely kill the node (and as such allow identifying users who visit a freesite by simply checking which nodes drop out of the network after sending a link to a freesite in a forum).

File: src/freenet/client/filter/GenericReadFilterCallback.java [hunk 1]
```diff
@@ -177,7 +178,7 @@ public String processURI(String u, String overrideType, boolean forBaseHref, boo
 				// @see bug #2297
 				return path;
 			} else if (linkFilterExceptionProvider != null) {
-				if (linkFilterExceptionProvider.isLinkExcepted(uri)) {
+				if (linkFilterExceptionProvider.isLinkExcepted(uri) || path.equals(ImageCreatorToadlet.ROOT_URL)) {
```
Comment (Bombe) [inline src/freenet/client/filter/GenericReadFilterCallback.java:181]: This is pretty much what the `LinkFilterExceptedToadlet` is about. Implement that in the `ImageCreatorToadlet` and have the method return `true` for paths that you want to be exposable.

**Comment by Bombe** [comment]
Also, this will allow trivial DoS attacks from Freesites by not restricting `width` and `height`. Make those numbers very large and you a) will consume _a lot_ of memory and b) will use a lot of CPU trying to find the maximum font size (because the naïve implementation is O(n) while it probably could be O(ln n)).

**Comment by Bombe** [comment]
Also, empty strings and strings containing unprintable characters only will essentially block the thread forever.

Oh, and also: _why?_ I mean, having something that generates an image that you have almost zero control over (except by trial and error) doesn’t seem particularly useful to me.

**Comment by Olezha** [comment]
> If I understand this correctly, it accesses the ImageCreatorToadlet, and that needs reasonable upper limits for the width and height.

Yes, and also here is the assumption that this Toadlet for placeholders. Perhaps there may be another purpose?

> Oh, and also: _why?_ I mean, having something that generates an image that you have almost zero control over (except by trial and error) doesn’t seem particularly useful to me.

Such a niche exists and is in demand. However, I agree with the question and see that it is necessary either to fix the flaws and open access, or slightly reduce the amount of code.

**Comment by Bombe** [comment]
> Yes, and also here is the assumption that this Toadlet for placeholders. Perhaps there may be another purpose?

To be honest, I didn’t even know we had such a thing.

> > Oh, and also: _why?_ I mean, having something that generates an image that you have almost zero control over (except by trial and error) doesn’t seem particularly useful to me.

> Such a niche exists and is in demand. However, I agree with the question and see that it is necessary either to fix the flaws and open access, or slightly reduce the amount of code.
Does it? I would almost be surprised if you found 10 people _on this planet_ who knew that thing even existed.

I don’t think we should open this to the world by allowing it to exempt the filter at all but if we do it needs to be locked down tight before that happens.

**Comment by Olezha** [comment]
> Does it? I would almost be surprised if you found 10 people _on this planet_ who knew that thing even existed.

Yes, in another world there are people who in a search engine ask something like this: image placeholders

> I don’t think we should open this to the world by allowing it to exempt the filter at all but if we do it needs to be locked down tight before that happens.

In general, I agree and then we need to go the second way.

Despite the fact that there may be a load, restrictions on one request have been added. Further, these are already questions not to this Toadlet, but to the Executor.

**Comment by Bombe** [comment]
> > Does it? I would almost be surprised if you found 10 people _on this planet_ who knew that thing even existed.
> 
> Yes, in another world there are people who in a search engine ask something like this: image placeholders

I’m not talking about placeholder images, I’m talking about the functionality in Fred. I have serious doubts that this is mentioned _anywhere_ in either our documentation or user-generated documentation in Freenet which is why I believe nobody knows about this or actually uses it.

> Further, these are already questions not to this Toadlet, but to the Executor.

What questions? What executor?

File: src/freenet/clients/http/ImageCreatorToadlet.java [hunk 2]
```diff
@@ -60,51 +65,72 @@ public void handleMethodGET(URI uri, HTTPRequest req, ToadletContext ctx) throws
...
+		return ROOT_URL;
+	}
+
+	@Override
+	public boolean isLinkExcepted(URI link) {
+		return ROOT_URL.equals(link.getPath());
+	}
+
+	private void specifyMaximumFontSizeThatFitsInImage(Graphics2D g2, FontRenderContext fc,
+													   int imageWidth, int imageHeight, String text) {
```
[the remark this answers is not in the document]
Comment (ArneBab) [reply src/freenet/clients/http/ImageCreatorToadlet.java:112]: AFAIK the last decision (I disagreed with) was tabs for indentation, spaces for alignment.

**Comment by ArneBab** [comment]
I would rather guess that the imageCreator exists not for Freesites but for Plugins.

**Comment by Bombe** [comment]
> I would rather guess that the imageCreator exists not for Freesites but for Plugins.

The only place it’s used from is somewhere inside the `ajaxpush` package…

File: src/freenet/clients/http/ImageCreatorToadlet.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Olezha) [reply src/freenet/clients/http/ImageCreatorToadlet.java]: > 2.3.1 Whitespace characters
> 2. Tab characters are **not** used for indentation.

However, this is a rare case when it is permissible to mix tabs and spaces. Imho  
Another acceptable solution is to remove spaces from this line. Now i have done so

**Comment by Olezha** [comment]
> What questions? What executor?

If we open access to this Toadlet and some Freesite will contain too many of these pictures, then an increased load on the node may appear which will open such a Freesite. I realized that such a problem could be. Protection from such a problem, I think, should be outside this Toadlet.

**Comment by Bombe** [comment]
I still haven’t heard a compelling reason why we should open access to this toadlet to freesites in the first place. In hope of getting Freenet into a state where we can actually work with it (refactoring and making things better) I would very much argue that we need to _reduce_ Fred’s public and semi-public APIs as much as possible, and this PR does just the opposite with no real gain.

**Comment by Olezha** [comment]
You're right.

File: src/freenet/clients/http/ImageCreatorToadlet.java [hunk 3]
```diff
@@ -60,51 +64,67 @@ public void handleMethodGET(URI uri, HTTPRequest req, ToadletContext ctx) throws
...
 		}
 	}
 
 	@Override
 	public String path() {
-		return "/imagecreator/";
+		return ROOT_URL;
 	}
 
+	public void specifyMaximumFontSizeThatFitsInImage(Graphics2D g2, FontRenderContext fc,
```
Comment (Bombe) [inline src/freenet/clients/http/ImageCreatorToadlet.java:105]: Okay, I think once you turn this into a package-private method (so you can reach it from the test) it should be okay to merge… I really don’t want to enlarge Fred’s API surface more than necessary. :)

**Comment by ArneBab** [comment]
@Olezha can you do the final change requested by @Bombe (make specifyMaximumFontSizeThatFitsInImage private)?

**Comment by ArneBab** [comment]
@Bombe is this good to go now?

**Comment by Bombe** [comment]
Yeah, go for it, it’s not more public than before… :)

**Pull request merged**